### ISSUE TITLE
General Planning

### ISSUE THREAD

**Issue by cczhu** [CityofToronto/bdit_traffic_prophet#2]
Overview of the procedure to transform the TEPs-I codebase to something easily integrated into the Big Data Innovation Team's workflow and data pipelines.

The overall plan is to first get a handle for how the code is run, and what sorts of outputs it produces, then figure out how to integrate it into our systems (mainly I/O and runtime environment considerations), then rewrite the MATLAB components into Python.  We will work from the outside in, first considering the I/O and codebase for KCOUNT and Local-SVR, before moving to PRTCS.  PECOUNT-I and II will be considered last (especially since they're FORTRAN based).

Weeks are 5 days.

Whether a task can (or should) be spun off as a separate project is indicated in italics.

Essentials - will let us use the code and represents a minimal code handover from UTTRI:
- Locally build and run TEPs-I.  See if we can exactly reproduce previous outputs.  Timescale: 1 day.
- Speak with Arman about the overall code structure and any ambiguous details.  Timescale: 1 day.
- Determine I/O formats for each stage.  Write a Python script to dump Postgres DB values csv for model ingestion.  Determine how to alter data ingestion (eg. how to remove a count from being considered). 
 Time: 2 days.
- Create a set of model internal consistency checks.  These should simply read in the intermediate and final outputs of the model and compare against whatever Arman has already outputted.  (Eventually we'll want to do something more intelligent as an actual test suite).  This allows us to check if any replacement code components alter code outputs.  Timescale: 1 week.
- Create a set of validation tests (maybe in a Jupyter notebook) that check whether model outputs are sensible.  @aharpalaniTO already has plotting software in R for validation.  Timescale: 2 weeks.
- Total timescale: 4 weeks.

Fundamental Development - simple modifications to check/extend the model:
- Rename TEPS-I and its submodules to something more descriptive.  *Should be crowdsourced.* Timescale: 0.5 days.
- Replace MATLAB and R scripts with Python ones, calling FORTRAN scripts using `os` as needed.  This will enable us to begin modifying PRTCS, KCOUNT and Local-SVR.  Validate that the Python model returns same results.  Timescale: 6 weeks, but could be x2 longer if custom regression algorithms are needed.
  - PRTCS will need custom functions to calculate nearest permanent count stations to all short-term stations (extracting from Arman's zipped distance matrix) and determining the maximum a posteriori permanent count station to each short-term station.
  - KCOUNT might be rewritten using [PyKridge](https://pykrige.readthedocs.io), which works on top of scikit-learn.  Variogram models might instead use [scikit-gstat](https://scikit-gstat.readthedocs.io). 
 KCOUNT also needs shortest network distance (probably the same one from PRTCS).
  - Local-SVR might use [scikit-learn's implementation](https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVR.html).
- `f2py` FORTRAN modules, and perform validation that outputs are unchanged.  Timescale: 1 week.
- Determine a set of global goodness of fit metrics to judge models by.  Ideally we'd have a set of metrics for each step of the model, but we should start with the outputs of KCOUNT and Local-SVR.  Timescale: 2 days.
- Create a way to flag counts as bad.  Timescale: 1 week.
  - Determine simple heuristics for flagging a count as bad.  Re-train model on only good counts to see if it improves accuracy.  *Should be spun off.*  Timescale: 4 weeks.
- Basic Readthedocs to describe how to deploy, train, predict with and modify the model.  Timescale: 2 weeks.
- Total timescale: 11-15 weeks.

Advanced Development - involved, extensive modifications that will substantially improve the model:
- Devise procedures for calculating all raw inputs to model, including network distances (both between street segments and counting stations).  Use revised inputs on model and see if it improves accuracy.  Timescale: 4 weeks.
  - Network distances could be calculated in PostGIS, networkx, graph-tool or [pandana](https://udst.github.io/pandana/introduction.html#the-general-workflow).  *Should be spun off.*
- Currently, the code divides traffic directions into "positive" and "negative" - which [delineate whether traffic is going north-east or south-west](https://github.com/CityofToronto/bdit_volumes/tree/master/volume_project/flow_data_processing#grid-angle-correction) - and treats each as a separate network, which doesn't make much sense.  Develop a more sensible count station relationship algorithm, and deploy into PRTCS.  Timescale: 2 weeks.
- Rewrite [AFRIMA](https://en.wikipedia.org/wiki/Autoregressive_fractionally_integrated_moving_average) and spatiotemporal disaggregators in PECOUNT into Cython, and validate against the original.  Performance-test throughout the process to determine if Cythonization improves performance.  Timescale: 3 weeks.
- Convert Python scripts into submodules controlled by a workflow package such as [Luigi](https://github.com/spotify/luigi) or [joblib](https://dask-ml.readthedocs.io/en/latest/joblib.html).  Timescale: 4 weeks.
- Create hooks for additional sources of data (eg. intersection turning counts, SCOOT) to be used as training or validation data.  Test improvement to model accuracy when these sources are incorporated.  This is especially interesting because count stations with enough data to estimate AADT are heavily biased toward highways right now.  Timescale: 4 weeks, but could be longer if outputs significantly change.
  - Actually transforming SCOOT data into traffic counts is an ML project on its own.  *Can be spun off.*
- Consider additional inputs to KCOUNT and Local-SVR: currently values such as land use, population density and number of lanes are used, but we have access to and can produce additional features.  Timescale: 4 weeks?  *Can be spun off.*
- Brief primer and links to relevant resources for all algorithms in documentation, since we can't upload the papers themselves.  Could go in a Readthedocs.  Timescale: 3 weeks?
- Design documentation, possibly similar to the [Move Developer Handbook](https://www.notion.so/182de05ad8a94888b52ccc68093a497a#8da00ef06ab744bbafa1ca2a5b29ff6f).  Timescale: 3 weeks?
- Optimizing the code for speed and multithreading it, possibly using [dask](https://examples.dask.org/machine-learning.html).  Timescale: 2 weeks?
- Consider replacing the algorithms used in PRTCS, KCOUNT, Local-SVR.  Of the three, Local-SVR has the poorest predictive accuracy (Fig. 11 of the paper) and deserves the most attention (it's also the one with the least amount of data, so could benefit from additional sources).  This is partly based on my suspicion that an ANN/kridging hybrid (see [this](https://www.mdpi.com/2072-4292/10/12/1864) or [this](https://ieeexplore.ieee.org/document/8721554)) might return higher accuracy that just kridging.  See also [this comparison](https://www.sciencedirect.com/science/article/pii/S2211675315000482#f000015) and [this review](https://ieeexplore.ieee.org/document/8490353).  Timescale: 12 weeks?  **Can be spun off.**

Speculative Work:
- [Statsmodels](https://github.com/statsmodels/statsmodels) [doesn't have an ARFIMA model](https://github.com/statsmodels/statsmodels/issues/5054), but in doing this project we may elect or be forced to write a Python version (possibly even based off of `statsmodels.tsa`).  This would be an opportunity to contribute to a widely used codebase, though implementation may deviate from the project's goals.  Timescale: 3 weeks (depends on how much of the Cythonized AFRIMA code is useful)?
- Is it possible to replace PRTCS, KCOUNT and Local-SVR with a single framework or ensemble of ML models?  Timescale: ???

In addition to the times listed, I'll need some time to reproduce the math behind the model's algorithms, and experiment with them with toy data.  Tasks under Bare Essentials might also take longer than expected because of my unfamiliarity with the code, while tasks under High-Hanging Fruit might go faster because there are existing packages and solutions available.

**Comment by cczhu** [comment]
Since [no plan survives contact with the enemy](https://en.wikiquote.org/wiki/Helmuth_von_Moltke_the_Elder), locally building and running TEPs-I will probably take several weeks, as the code is not running, the codebase is quite complex and it is difficult to determine which intermediate steps are suitable for consistency checks.

Revising estimate for step 1 to 2 weeks, dependent on when Arman is able to respond to our questions.  See #4.

**Comment by cczhu** [comment]
Moving all planning work into the project Kanban.  The above checklist of resources and ideas is still useful, but not being followed as a work plan, so closing.
